prov/verbs/src/fi_verbs.c

```diff
@@ -2403,18 +2405,33 @@ fi_ibv_poll_events(struct fi_ibv_cq *_cq, int timeout)
 {
 	int ret;
 	void *context;
+	struct pollfd fds[2];
+	char data;
 
-	ret = fi_poll_fd(_cq->channel->fd, timeout);
+	fds[0].fd = _cq->channel->fd;
+	fds[1].fd = _cq->signal_fd[0];
+
+	fds[0].events = fds[1].events = POLLIN;
+
+	ret = poll(fds, 2, timeout);
 	if (ret == 0)
 		return -FI_EAGAIN;
 	else if (ret < 0)
-		return ret;
+		return -errno;
 
-	ret = ibv_get_cq_event(_cq->channel, &_cq->cq, &context);
-	if (ret)
-		return ret;
+	if (fds[1].revents & POLLIN) {
+		read(fds[1].fd, &data, 1);
```

Comment:
To be safe, I would call read in a loop until it returns no data.  This fixes a race where signal is either called twice or signal is called immediately after a waiting thread times out.  Either case can result in the next call getting the signal and returning immediately.